### PR TITLE
[Azure Search] Add functional tests for custom analyzers

### PR DESCRIPTION
**Pull request by loic-sharma**
Adds functional tests for custom analyzers added in https://github.com/NuGet/NuGet.Services.Metadata/pull/485. Note that these functional tests do not run on deployments, that is tracked by: https://github.com/NuGet/NuGetGallery/issues/6962

### CODE REVIEW

File: tests/AzureSearchTests.FunctionalTests/AzureIndexFunctionalTests.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using BasicSearchTests.FunctionalTests.Core;
+using Newtonsoft.Json;
+
+namespace AzureSearchTests.FunctionalTests
```
Comment (joelverhagen) [inline tests/AzureSearchTests.FunctionalTests/AzureIndexFunctionalTests.cs]: Namespace should be similar to the library under test.

File: tests/AzureSearchTests.FunctionalTests/AzureSearchTests.FunctionalTests.csproj [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    </PackageReference>
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\BasicSearchTests.FunctionalTests.Core\BasicSearchTests.FunctionalTests.Core.csproj">
+      <Project>{eea7b6c1-0358-4e67-9d2a-e30b8ff9ff3d}</Project>
+      <Name>BasicSearchTests.FunctionalTests.Core</Name>
+    </ProjectReference>
+  </ItemGroup>
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <!-- To modify your build process, add your task inside one of the targets below and uncomment it.
```
Comment (joelverhagen) [inline tests/AzureSearchTests.FunctionalTests/AzureSearchTests.FunctionalTests.csproj]: nit: delete unnecessary stuff

File: tests/AzureSearchTests.FunctionalTests/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+[assembly: AssemblyTitle("AzureSearchTests.FunctionalTests")]
```
Comment (joelverhagen) [inline tests/AzureSearchTests.FunctionalTests/Properties/AssemblyInfo.cs]: nit: delete unnecessary stuff

File: tests/BasicSearchTests.FunctionalTests.Core/BaseFunctionalTests.cs [hunk 4]
```diff
@@ -14,11 +14,16 @@ public class BaseFunctionalTests
...
+        {
+        }
+
+        public BaseFunctionalTests(string baseUrl)
         {
             // Arrange
             IgnoreCertificateValidationErrors();
             RetryHandler = new RetryHandler(new HttpClientHandler());
-            Client = new HttpClient(RetryHandler) { BaseAddress = new Uri(EnvironmentSettings.SearchServiceBaseUrl) };
+            Client = new HttpClient(RetryHandler) { BaseAddress = new Uri(baseUrl) };
```
Comment (joelverhagen) [inline tests/BasicSearchTests.FunctionalTests.Core/BaseFunctionalTests.cs]: `Client` should be disposed by making the test class `IDisposable`. I think `RetryHandler` too

File: tests/AzureSearchTests.FunctionalTests/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+using System.Reflection;
```
Comment (joelverhagen) [inline tests/AzureSearchTests.FunctionalTests/Properties/AssemblyInfo.cs]: nit: copyright header

File: tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs [hunk 6]
```diff
@@ -47,6 +50,45 @@ public static string IndexBaseUrl
             }
         }
 
+        public static string AzureSearchIndexUrl
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(_azureSearchIndexUrl))
+                {
+                    _azureSearchIndexUrl = GetEnvironmentVariable("AzureSearchIndexUrl", "https://nuget-int-usnc.search.windows.net");
```
Comment (joelverhagen) [inline tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs]: Can we throw if the variable is not defined? Defaulting to INT seems pretty odd.

File: tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs [hunk 7]
```diff
@@ -47,6 +50,45 @@ public static string IndexBaseUrl
...
+                return _azureSearchIndexUrl;
+            }
+        }
+
+        public static string AzureSearchIndexApiKey
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(_azureSearchIndexApiKey))
+                {
```
Comment (joelverhagen) [inline tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs]: throw?

File: tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs [hunk 8]
```diff
@@ -47,6 +50,45 @@ public static string IndexBaseUrl
...
+            }
+        }
+
+        public static string AzureSearchIndexName
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(_azureSearchIndexName))
+                {
+                    _azureSearchIndexName = GetEnvironmentVariable("AzureSearchIndexName", "search");
```
Comment (joelverhagen) [inline tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs]: throw?

File: tests/AzureSearchTests.FunctionalTests/AzureIndexFunctionalTests.cs (hunk 1 above)
Comment (loic-sharma) [reply tests/AzureSearchTests.FunctionalTests/AzureIndexFunctionalTests.cs]: Fixed

File: tests/AzureSearchTests.FunctionalTests/AzureSearchTests.FunctionalTests.csproj (hunk 2 above)
Comment (loic-sharma) [reply tests/AzureSearchTests.FunctionalTests/AzureSearchTests.FunctionalTests.csproj]: Removed

File: tests/BasicSearchTests.FunctionalTests.Core/BaseFunctionalTests.cs (hunk 4 above)
Comment (loic-sharma) [reply tests/BasicSearchTests.FunctionalTests.Core/BaseFunctionalTests.cs]: Disposed both `Client` and `RetryHandler`

File: tests/AzureSearchTests.FunctionalTests/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (loic-sharma) [reply tests/AzureSearchTests.FunctionalTests/Properties/AssemblyInfo.cs]: Added

File: tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs (hunk 6 above)
Comment (loic-sharma) [reply tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs]: Talked offline. This is the pattern used by other settings, I will keep this setting consistent.

File: tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs (hunk 7 above)
Comment (loic-sharma) [reply tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs]: Made this a throw as this is a secret

File: tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs (hunk 8 above)
Comment (loic-sharma) [reply tests/BasicSearchTests.FunctionalTests.Core/EnvironmentSettings.cs]: Talked offline. This is the pattern used by other settings, I will keep this setting consistent.

File: tests/AzureSearchTests.FunctionalTests/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (loic-sharma) [reply tests/AzureSearchTests.FunctionalTests/Properties/AssemblyInfo.cs]: Deleted

**Pull request merged**